### PR TITLE
More `shelleyBasedEraConstraint` constraints

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    More shelleyBasedEraConstraint constraints
  compatibility: compatible
  type: feature
```

# Context

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/Orphans.hs [hunk 1]
```diff
@@ -1,20 +1,46 @@
...
 {-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Orphans.hs:12]: Can we avoid this extension? Why it's needed?

**Review by carbolymer — CHANGES_REQUESTED** [review]
https://github.com/input-output-hk/cardano-api/pull/149/files#r1277303412

File: cardano-api/internal/Cardano/Api/Orphans.hs (hunk 1 above)
Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/Orphans.hs:12]: This is why we need it:

```

internal/Cardano/Api/Orphans.hs:97:10: error:
    • Illegal nested constraint ‘ToJSON (HeaderHash blk)’
      (Use UndecidableInstances to permit this)
    • In the instance declaration for ‘ToJSON (Tip blk)’
   |
97 | instance ToJSON (HeaderHash blk) => ToJSON (Tip blk) where
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Note, the offending instance already exists in `cardano-cli`.  This PR merely moves it here.

Comment (carbolymer) [reply cardano-api/internal/Cardano/Api/Orphans.hs:12]: Your link returns 404 error for me.

**Review by carbolymer — APPROVED** [review]

**Pull request merged**